Initial testing docs

benrudolph opened the pull request:
@dimagi/test-team first go at the docs, please feel free to add commits

Commit:
Initial testing docs

Commit:
fix link

benrudolph commented:
hehe killed build

also didn't add anything about travis. we may want a section in there

emord commented:
Haven't thought about this in depth yet, but maybe we should have a list of common mocks that one can use. I just found out about the `UserESFake` last week and it could have been useful in the past

emord commented on docs/testing.rst:
```diff
@@ -0,0 +1,60 @@
+
+.. code:: python
+
+    class MyBaseTestClass(TestCase):
```
This might be a bad example. I think that you would want the base class to inherit from object and have the child classes inherit tow classes http://stackoverflow.com/questions/4566910/abstract-test-case-using-python-unittest

benrudolph replied:
ah i think this is supposed to be a bad example, or are you saying that this is a bad "bad example". that said maybe we shouldn't put bad examples in it

benrudolph replied:
trying to make an example out of what happened in ilsgateway

emord replied:
Oh. Gotcha that makes sense. Although these setups would actually run 3 times because  all three inherit from `TestCase`

millerdev replied:
I'm pro not putting bad examples in docs without clearly marking them with something like
```
# BAD CODE; DO NOT DO THIS
```
Also might be worth putting an example of the right way to do it.

millerdev commented on docs/testing.rst:
```diff
@@ -0,0 +1,60 @@
+Tearing down tests
+==================
+
+It is good practice to ensure that all objects you have created in the test database are deleted when the test
```
Nit: "It is good practice to ensure..." sounds like it's good but not required. Suggest changing to "It is important to ensure..."

emord commented on docs/testing.rst:
```diff
@@ -0,0 +1,60 @@
+Using setUpClass
+================
+
+Doing a lot of work in the ``setUp`` call of a test class means that it will be run on every test. This
```
There's also package and module. http://pythontesting.net/framework/nose/nose-fixture-reference/#package

Commit:
PR feedback

benrudolph commented:
thanks for the feedback guys. i've updated with all your comments. emord i added a small list of mock libraries HQ has, though i'm not sure if it's exhaustive

millerdev commented on docs/testing.rst:
```diff
@@ -0,0 +1,98 @@
+
+Sometimes classes share the same base class and inherit the ``setUpClass`` function. Below is an example:
+
+.. code:: python
```
These code samples are not showing up in the [rich text view](https://github.com/dimagi/commcare-hq/blob/4bed22abf2f22b582d9af0d4ba63cf1da31e9e08/docs/testing.rst)

benrudolph replied:
👍 will take a look, thought it was working. thanks for checking

millerdev commented on docs/testing.rst:
```diff
@@ -0,0 +1,98 @@
+            ...
+
+
+In the above example the ``setUpClass`` is run three times, once for ``MyBaseTestClass``, ``MyTestClass`` and ``MyTestClassTwo``. If
```
> ``setUpClass`` is run three times

In `BAD EXAMPLE`, this is not true, it's only run twice because `MyBaseTestClass` has no test methods.

Also, you say "In the above example" but both GOOD and BAD examples are above and it's not clear which you're referring to, and the GOOD example is above the BAD example, so that adds to the confusion.

benrudolph replied:
👍 sorry i had originally thought that it was only run twice too, but then changed based on this comment: https://github.com/dimagi/commcare-hq/pull/15122#discussion_r102724606. i should just run it and figure out myself

millerdev replied:
Did that:
```py
from unittest import TestCase

class MyBaseTestClass(TestCase):

    runs = 0

    @classmethod
    def setUpClass(cls):
        MyBaseTestClass.runs += 1
        print("setup", MyBaseTestClass.runs)

    def _my_helper_function(self):
        print("help")


class MyTestClass(MyBaseTestClass):

    def test(self):
        pass

class MyTestClassTwo(MyBaseTestClass):

    def test(self):
        pass
```
output:
```
$ nosetests sampletest.py -vs
('setup', 1)
test (sampletest.MyTestClass) ... ok
('setup', 2)
test (sampletest.MyTestClassTwo) ... ok

----------------------------------------------------------------------
Ran 2 tests in 0.001s

OK
```

millerdev commented on docs/testing.rst:
```diff
@@ -0,0 +1,98 @@
+
+
+In the above example the ``setUpClass`` is run three times, once for ``MyBaseTestClass``, ``MyTestClass`` and ``MyTestClassTwo``. If
+``setUpClass`` has expensive operations, then it's best for all the tests to be combined under one test class.
```
Nit: this feels like a separate point, and isn't really relevant to the examples given since neither GOOD nor BAD examples are doing what this is saying: combine all tests under one test class. Maybe put it in a separate paragraph?

Commit:
Restructure testing docs

Commit:
Better wording

millerdev commented on docs/testing.rst:
```diff
@@ -0,0 +1,97 @@
+            ...
+
+
+In the above example the ``setUpClass`` is run twice, once for ``MyTestClass`` and once for ``MyTestClassTwo``. If ``setUpClass`` has expensive operations, then it's best for all the tests to be combined under one test class.
```
I hate to keep beating this horse, but the example below does not combine all tests "under one test class" as this sentence describes. It has two test classes with separate `setUpClass` methods.

This code example appears to be demonstrating how to share a helper method with multiple test classes using a mixin, which is a good technique to document. However, it seems like a separate point from merging tests to avoid running expensive setup repeatedly.

benrudolph replied:
Please keep beating the horse haha (sorry I haven't had the time I wanted to sit down and write this cleanly and appreciate this review. i'll ping when i think ready for another review, didn't mean to make you review again)

benrudolph replied:
also i do see what you mean, i had initially thought your comment was pointing to another issue

benrudolph replied:
https://github.com/dimagi/commcare-hq/pull/15122/commits/75c4346ab188651ce3f727dc7edfd9c80327c69d -- ok i think this example is better

Commit:
better examle

Commit:
better bullet list

millerdev reviewed (approved):
Not sure if you're done with this yet, but looks great!

benrudolph commented:
for now, will add more incrementally 📈
